apps/arweave/src/ar_p3.erl

```diff
@@ -0,0 +1,325 @@
+-module(ar_p3).
+
+-behaviour(gen_server).
+
+-include_lib("arweave/include/ar.hrl").
+-include_lib("arweave/include/ar_config.hrl").
+-include_lib("arweave/include/ar_p3.hrl").
+
+-export([start_link/0, allow_request/1, get_balance/3, get_rates_json/0]).
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).
+
+-ifdef(TEST).
+-define(MAX_BLOCK_SCAN, 4).
+-else.
+-define(MAX_BLOCK_SCAN, 200).
+-endif.
+
+%%%===================================================================
+%%% Public interface.
+%%%===================================================================
+allow_request(Req) ->
+	gen_server:call(?MODULE, {request, Req}).
```

Comment:
I think, the handler might very occasionally take longer than 5 seconds to execute so I would catch the exit and return a timeout error explicitly.

Reply:
will do! Thanks.